core/src/main/java/io/questdb/griffin/SqlCodeGenerator.java

```diff
@@ -2095,62 +2096,87 @@ private RecordCursorFactory generateTableQuery(
+                                } else {
+                                    rcf = new DeferredSymbolIndexFilteredRowCursorFactory(keyColumnIndex, Chars.toString(symbol), f, true, indexDirection);
+                                }
                             } else {
-                                rcf = new DeferredSymbolIndexFilteredRowCursorFactory(keyColumnIndex, Chars.toString(symbol), f, true, indexDirection);
+                                if (f == null) {
+                                    rcf = new SymbolIndexRowCursorFactory(keyColumnIndex, symbolKey, true, indexDirection);
+                                } else {
+                                    rcf = new SymbolIndexFilteredRowCursorFactory(keyColumnIndex, symbolKey, f, true, indexDirection);
+                                }
                             }
-                        } else {
-                            if (f == null) {
-                                rcf = new SymbolIndexRowCursorFactory(keyColumnIndex, symbolKey, true, indexDirection);
+                            return new DataFrameRecordCursorFactory(myMeta, dfcFactory, rcf, orderByKeyColumn, f, false, columnIndexes, columnSizes);
+                        }
+
+                        symbolValueList.clear();
+
+                        for (int i = 0, n = intrinsicModel.keyValues.size(); i < n; i++) {
+                            symbolValueList.add(intrinsicModel.keyValues.get(i));
+                        }
+
+                        if (orderByKeyColumn) {
+                            myMeta.setTimestampIndex(-1);
+                            if (model.getOrderByDirectionAdvice().getQuick(0) == QueryModel.ORDER_DIRECTION_ASCENDING) {
+                                symbolValueList.sort(Chars.CHAR_SEQUENCE_COMPARATOR);
                             } else {
-                                rcf = new SymbolIndexFilteredRowCursorFactory(keyColumnIndex, symbolKey, f, true, indexDirection);
+                                symbolValueList.sort(Chars.CHAR_SEQUENCE_COMPARATOR_DESC);
                             }
                         }
-                        return new DataFrameRecordCursorFactory(myMeta, dfcFactory, rcf, orderByKeyColumn, f, false, columnIndexes, columnSizes);
-                    }
 
-                    symbolValueList.clear();
-
-                    for (int i = 0, n = intrinsicModel.keyValues.size(); i < n; i++) {
-                        symbolValueList.add(intrinsicModel.keyValues.get(i));
-                    }
+                        final Function f = compileFilter(intrinsicModel, readerMeta, executionContext);
+                        if (f != null && f.isConstant() && !f.getBool(null)) {
+                            return new EmptyTableRecordCursorFactory(myMeta);
+                        }
+                        return new FilterOnValuesRecordCursorFactory(
```

Comment:
I recently tackled potential memory leak in Functions. Function is Closeable interface.

When 'f' is constant we will be discarding it regardless if it is true or false. 'false' branch is ok, but 'true' branch needs to behave as if there is no filter. In both cases 'f' need to be closed. I usually do something like:

```java
if (f.isConstant()) {
   try (f) {
       if (f.getBool(null)) {
       } else {
       }
   }
}
```